mm2src/mm2_main/src/lp_swap/taker_swap.rs

```diff
@@ -1315,28 +1321,54 @@ impl TakerSwap {
 
         let mut swap_events = vec![TakerSwapEvent::TakerPaymentSent(tx_ident)];
         if self.ctx.use_watchers() {
-            let preimage_fut = self.taker_coin.create_taker_spends_maker_payment_preimage(
+            let taker_spends_maker_payment_fut = self.taker_coin.create_taker_spends_maker_payment_preimage(
                 &self.r().maker_payment.as_ref().unwrap().tx_hex,
                 self.maker_payment_lock.load(Ordering::Relaxed) as u32,
                 self.r().other_maker_coin_htlc_pub.as_slice(),
                 &self.r().secret_hash.0,
                 &self.unique_swap_data()[..],
             );
 
-            match preimage_fut.compat().await {
-                Ok(preimage) => {
-                    let watcher_data = self.create_watcher_data(transaction.tx_hex(), preimage.tx_hex());
-                    let swpmsg_watcher = SwapWatcherMsg::TakerSwapWatcherMsg(Box::new(watcher_data));
-                    broadcast_swap_message(
-                        &self.ctx,
-                        watcher_topic(&self.r().data.taker_coin),
-                        swpmsg_watcher,
-                        &self.p2p_privkey,
+            match taker_spends_maker_payment_fut.compat().await {
```

Comment:
This code block looks too nested. It should be possible to use [try_join](https://docs.rs/futures/latest/futures/future/fn.try_join.html) here.

Reply:
This line is related to the current comment
https://github.com/KomodoPlatform/atomicDEX-API/blob/watchtower/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L1367

As per [try_join docs](https://docs.rs/futures/latest/futures/future/fn.try_join.html),
> If unsuccessful, it will complete with the first error encountered.

It's not guaranteed that refund preimage generation was failed. `Failed to create at least one of preimages` will be a better match.